storagenode/piecestore/inspector.go

```diff
@@ -4,14 +4,120 @@
 package piecestore
 
 import (
+	"context"
+	"strings"
+	"time"
+
+	"storj.io/storj/storagenode/bandwidth"
+
+	"github.com/golang/protobuf/ptypes"
+	"go.uber.org/zap"
+
 	"storj.io/storj/pkg/kademlia"
-	"storj.io/storj/storagenode/orders"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/piecestore/psserver"
+	"storj.io/storj/pkg/storj"
 	"storj.io/storj/storagenode/pieces"
 )
 
 // Inspector does inspectory things
 type Inspector struct {
-	pieces   *pieces.Store
-	kademlia *kademlia.Kademlia
-	orders   orders.DB
+	log       *zap.Logger
+	pieceInfo pieces.DB
+	kademlia  *kademlia.Kademlia
+	usageDB   bandwidth.DB
+
+	startTime time.Time
+	config    psserver.Config
+}
+
+// NewInspector creates piecestore inspector instance
+func NewInspector(log *zap.Logger, pieceInfo pieces.DB, kademlia *kademlia.Kademlia, usageDB bandwidth.DB, config psserver.Config) *Inspector {
+	return &Inspector{
+		log:       log,
+		pieceInfo: pieceInfo,
+		kademlia:  kademlia,
+		usageDB:   usageDB,
+		config:    config,
+		startTime: time.Now(),
+	}
+}
+
+func (inspector *Inspector) retrieveStats(ctx context.Context) (*pb.StatSummaryResponse, error) {
+	totalUsedSpace, err := inspector.pieceInfo.SpaceUsed(ctx)
+	if err != nil {
+		return nil, err
+	}
+	usage, err := inspector.usageDB.Summary(ctx, getBeginningOfMonth(), time.Now())
+	if err != nil {
+		return nil, err
+	}
+
+	totalUsedBandwidth := usage.Total()
+
+	return &pb.StatSummaryResponse{
+		UsedSpace:          totalUsedSpace,
+		AvailableSpace:     (inspector.config.AllocatedDiskSpace.Int64() - totalUsedSpace),
+		UsedBandwidth:      totalUsedBandwidth,
+		AvailableBandwidth: (inspector.config.AllocatedBandwidth.Int64() - totalUsedBandwidth),
+	}, nil
+}
+
+// Stats returns current statistics about the storage node
+func (inspector *Inspector) Stats(ctx context.Context, in *pb.StatsRequest) (*pb.StatSummaryResponse, error) {
+	inspector.log.Debug("Getting Stats...")
+
+	statsSummary, err := inspector.retrieveStats(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	inspector.log.Info("Successfully retrieved Stats...")
+
+	return statsSummary, nil
+}
+
+func (inspector *Inspector) getDashboardData(ctx context.Context) (*pb.DashboardResponse, error) {
+	statsSummary, err := inspector.retrieveStats(ctx)
+	if err != nil {
+		return &pb.DashboardResponse{}, Error.Wrap(err)
+	}
+
+	nodes, err := inspector.kademlia.FindNear(ctx, storj.NodeID{}, 10000000)
```

Comment:
I wonder what would be the performance impact of asking Kademlia to find the nearest 10 million nodes in a large network?

Reply:
I guess, not great.

Reply:
We have also `kademlia.DumpNodes` that returns all noes from the local routing table.

If the goal is just to count the number of nodes in the routing table, then perhaps it would be better to add `kademlia.NodeCount` method that will make this in a more optimal way, without wasting memory, but just increasing a counter.

Reply:
`DumpNodes` is used for debugging kademlia. Just remembered that, `FindNear` looks up from local database and it doesn't actually query the network.

Reply:
But we just want to count the number of nodes in the routing table (i.e. the neighborhood size), shouldn't we have a more efficient way? Remember, the dashboard will call this every 3 seconds along with other things.

Reply:
Ah, right... yeah. We can improve that later I guess.

Reply:
Sure. Let's just have a TODO comment for now.

Reply:
@kaloyan-raev can you add TODO?

Reply:
added